torch/distributed/rpc/api.py

```diff
@@ -66,10 +90,42 @@ def _wait_all_workers():
     terminate the RPC framework, and there is no guarantee that the RPC
     framework will work after this method returns.
     """
-    global _agent
+    if _agent is None:
+        return
+
+    assert _ALL_WORKER_NAMES is not None, (
+        "`_ALL_WORKER_NAMES` is not initialized for `def _wait_all_workers`."
+    )
+    leader_worker_name = sorted(_ALL_WORKER_NAMES)[0]
+
+    self_worker_name = _agent.get_worker_info().name
+    assert self_worker_name not in _INTENDED_WORKER_NAMES, (
+        "Can not call _wait_all_workers() twice."
+    )
+
+    is_leader_worker = leader_worker_name == self_worker_name
+
+    # Phase 1: Followers send intents.
+    # All followers report intents to the leader.
+    if is_leader_worker:
+        _on_leader_follower_report_intent(self_worker_name)
+    else:
+        rpc_async(
+            leader_worker_name,
+            _on_leader_follower_report_intent,
+            args=(self_worker_name,),
+        )
+
+    _PROCEED_SIGNAL.wait()
+
+    # Phase 2: Leader asks followers to proceed.
+    # Leader's signal is the first to be unblocked,
+    # after receiving all followers' intents.
+    if is_leader_worker:
+        # The leader sends out proceeed signals to all followers.
+        for follower_worker_name in _ALL_WORKER_NAMES - {leader_worker_name}:
+            rpc_async(follower_worker_name, _set_proceed_signal, args=())
```

Comment:
After this line the leader will do a local shutdown and it may lead to _set_proceed_signal not go out to the nodes as the local agent may be killed. What can we do here to avoid this other than making leader wait for an arbitrary time ?